replication-api-impl/src/main/java/org/codice/ditto/replication/api/impl/Syncer.java

```diff
@@ -0,0 +1,342 @@
+package org.codice.ditto.replication.api.impl;
+
+import java.util.Collections;
+import java.util.Date;
+import java.util.List;
+import java.util.Optional;
+import javax.annotation.Nullable;
+import javax.ws.rs.NotFoundException;
+import org.codice.ditto.replication.api.NodeAdapter;
+import org.codice.ditto.replication.api.Replication;
+import org.codice.ditto.replication.api.ReplicationItem;
+import org.codice.ditto.replication.api.Status;
+import org.codice.ditto.replication.api.data.Metadata;
+import org.codice.ditto.replication.api.data.QueryRequest;
+import org.codice.ditto.replication.api.data.ReplicationStatus;
+import org.codice.ditto.replication.api.data.ReplicatorConfig;
+import org.codice.ditto.replication.api.data.Resource;
+import org.codice.ditto.replication.api.data.ResourceResponse;
+import org.codice.ditto.replication.api.impl.data.CreateRequestImpl;
+import org.codice.ditto.replication.api.impl.data.CreateStorageRequestImpl;
+import org.codice.ditto.replication.api.impl.data.DeleteRequestImpl;
+import org.codice.ditto.replication.api.impl.data.QueryRequestImpl;
+import org.codice.ditto.replication.api.impl.data.ResourceRequestImpl;
+import org.codice.ditto.replication.api.impl.data.UpdateRequestImpl;
+import org.codice.ditto.replication.api.impl.data.UpdateStorageRequestImpl;
+import org.codice.ditto.replication.api.persistence.ReplicationItemManager;
+import org.codice.ditto.replication.api.persistence.ReplicatorHistoryManager;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Performs creates, updates and deletes between source and destination {@link NodeAdapter}s based
+ * on the given {@link ReplicatorConfig}.
+ */
+public class Syncer {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(Syncer.class);
+
+  private final ReplicationItemManager replicationItemManager;
+
+  private final ReplicatorHistoryManager historyManager;
+
+  public Syncer(
+      ReplicationItemManager replicationItemManager, ReplicatorHistoryManager historyManager) {
+    this.replicationItemManager = replicationItemManager;
+    this.historyManager = historyManager;
+  }
+
+  public Job create(
+      NodeAdapter source,
+      NodeAdapter destination,
+      ReplicatorConfig replicatorConfig,
+      ReplicationStatus replicationStatus) {
+    return new Job(source, destination, replicatorConfig, replicationStatus);
+  }
+
+  class Job {
+
+    private final NodeAdapter source;
+
+    private final NodeAdapter destination;
+
+    private final String sourceName;
+
+    private final String destinationName;
+
+    private final ReplicatorConfig replicatorConfig;
+
+    private ReplicationStatus replicationStatus;
```

Comment:
❗️ Because it will be accessed by multiple threads, special considerations applies to both the reference and the object itself. If the reference is never changed then we should make it final. All access to the object should be protected by synchronization or the object itself should be designed to be thread safe.

Reply:
I added a javadoc note. But this status object will not be accessed by multiple threads because it is a new one each time, which once the job finishes is compounded with any existing status objects in the database for the given config.